File: shop/automator/core/status.py
import logging
import time
from .api import logger
from .config import LIVE_DAYS
from .models import DROPPED, OPEN, USED, Batch, Task
from .s3 import STATUS_KEY, TASKS_PREFIX, get_json, list_batches, list_objects, load_strikes, put_json

LINES = 300
ERROR_LIMIT = 300

class Journal(logging.Handler):

    def __init__(self):
        super().__init__(level=logging.INFO)
        self.lines: list[str] = []

    def emit(self, record: logging.LogRecord) -> None:
        stamp = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime(record.created))
        self.lines.append(f"{stamp} {record.levelname} {record.getMessage()}")

def attach() -> Journal:
    journal = Journal()
    logger.addHandler(journal)
    return journal

def detach(journal: Journal) -> None:
    logger.removeHandler(journal)

def task_summary(task: Task) -> dict:
    if not task:
        return None
    metadata = dict(task.metadata)
    if metadata.get("failed_error"):
        metadata["failed_error"] = str(metadata["failed_error"])[:ERROR_LIMIT]
    return {
        "key": task.key,
        "step": task.step,
        "design": task.design,
        "product": {"id": task.product.id, "title": task.product.title, "category": task.product.category},
        "created_at": task.created_at,
        "updated_at": task.updated_at,
        "printfiles": len(task.printfiles),
        "mockups": len(task.mockups),
        "variants": len(task.variants),
        "metadata": metadata,
    }

def design_summary(batch: Batch) -> dict:
    if not batch:
        return None
    return {
        "key": batch.design,
        "created_at": batch.created_at,
        "group": batch.group,
        "edition": batch.variation.get("edition"),
        "secondary": batch.paint.get("secondary"),
        "scheme": batch.paint.get("scheme"),
    }

def batch_summary(batch: Batch) -> dict:
    if not batch:
        return None
    return {
        "open": len(batch.cells(OPEN)),
        "used": len(batch.cells(USED)),
        "dropped": len(batch.cells(DROPPED)),
        "tiles": batch.tiles,
        "strikes": load_strikes(),
    }

def live_summary(now: float) -> dict:
    keys = [obj["Key"] for obj in list_objects(TASKS_PREFIX)]
    batches = [batch for batch in list_batches() if batch.released_at]
    cutoff = now - LIVE_DAYS * 86400
    counts = {batch.design: sum(1 for key in keys if key.endswith(f"-{batch.design}.json")) for batch in batches}
    stamps = [batch.released_at for batch in batches]
    return {
        "products": len(keys),
        "batches": len(batches),
        "expiring": sum(counts[batch.design] for batch in batches if batch.released_at < cutoff + 86400),
        "newest": max(stamps) if stamps else None,
        "oldest": min(stamps) if stamps else None,
    }

def previous_log() -> list[str]:
    try:
        return list(get_json(STATUS_KEY).get("log") or [])
    except Exception:
        return []

def write(journal: Journal, task: Task, batch: Batch, seconds: float, error: str = None) -> None:
    now = time.time()
    log = (previous_log() + journal.lines)[-LINES:]
    data = {
        "at": int(now),
        "seconds": round(seconds, 1),
        "error": error,
        "design": design_summary(batch),
        "batch": batch_summary(batch),
        "task": task_summary(task),
        "live": live_summary(now),
        "log": log,
    }
    put_json(STATUS_KEY, data, cache="no-cache")

File: site/src/client/status.ts
const root = document.querySelector<HTMLElement>("[data-status]");

const EVERY = 60 * 1000;

type Automator = {
  at: number;
  seconds: number;
  error: string | null;
  design: { key: string; created_at: number; group: string; edition: string; primary: string; secondary: string[]; scheme: string } | null;
  task: { key: string; step: string; product: { title: string }; created_at: number; updated_at: number; mockups: number; variants: number; metadata: Record<string, unknown> } | null;
  paths: { open: number; used: number; quarantined: number; strikes?: Record<string, number> } | null;
  live: { products: number; expiring: number; newest: number | null; oldest: number | null };
  log: string[];
};

type Stats = {
  at: number;
  hours: number;
  cdn: { requests?: number; bytes?: number; error_4xx?: number; error_5xx?: number };
  lambdas: Record<string, { invocations: number; errors: number; throttles: number; duration_ms: number }>;
  errors: Record<string, { at: string | null; level: string; message: string }[]>;
  bucket: { site_objects: number; site_bytes: number; data_objects: number; data_bytes: number; products: number; designs: number; posts: number };
};

const when = (seconds: number | null | undefined) => (seconds ? new Date(seconds * 1000).toISOString().replace("T", " ").slice(0, 19) + " UTC" : "never");

const ago = (seconds: number | null | undefined) => {
  if (!seconds) return "never";
  const gap = Math.max(0, Math.floor(Date.now() / 1000 - seconds));
  if (gap < 90) return `${gap} s ago`;
  if (gap < 5400) return `${Math.round(gap / 60)} min ago`;
  if (gap < 172800) return `${(gap / 3600).toFixed(1)} h ago`;
  return `${Math.round(gap / 86400)} d ago`;
};

const bytes = (n: number | undefined) => (n == null ? "" : n < 1e6 ? `${(n / 1e3).toFixed(0)} kB` : n < 1e9 ? `${(n / 1e6).toFixed(1)} MB` : `${(n / 1e9).toFixed(2)} GB`);

const esc = (text: unknown) => String(text ?? "").replace(/[&<>"]/g, (c) => ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;" })[c] ?? c);

const rows = (list: [string, unknown][]) => `<table>${list.map(([name, value]) => `<tr><th>${esc(name)}</th><td>${esc(value)}</td></tr>`).join("")}</table>`;

const LINE = /^(\S+)\s+(DEBUG|INFO|WARNING|ERROR|CRITICAL)\s+(.*)$/;

const paint = (line: string) => {
  const hit = LINE.exec(line);
  if (!hit) return `<span class="text">${esc(line)}</span>`;
  const level = hit[2].toLowerCase();
  return `<span class="${level}"><span class="time">${esc(hit[1])}</span> <span class="level">${esc(hit[2].padEnd(7))}</span> <span class="text">${esc(hit[3])}</span></span>`;
};

const code = (lines: string[]) => `<pre><code>${lines.map(paint).join("\n")}</code></pre>`;

async function fetchJson<T>(path: string): Promise<T | null> {
  try {
    const reply = await fetch(path, { cache: "no-store" });
    return reply.ok ? ((await reply.json()) as T) : null;
  } catch {
    return null;
  }
}

function automator(data: Automator | null) {
  const box = root?.querySelector("[data-automator]");
  if (!box) return;
  if (!data) {
    box.innerHTML = "<h2>Automator</h2><p class=\"fine\">No status file yet.</p>";
    return;
  }
  const task = data.task;
  const waiting = task?.metadata?.waiting_since as number | undefined;
  const list: [string, unknown][] = [
    ["Last tick", `${ago(data.at)} (${when(data.at)}), ${data.seconds} s${data.error ? `, ${data.error}` : ""}`],
    ["Design", data.design ? `${data.design.key}, ${data.design.group} ${data.design.edition} ${data.design.scheme}, ${data.design.primary} with ${data.design.secondary.join(", ")}, ${ago(data.design.created_at)}` : "none"],
    ["Task", task ? `${task.key} at ${task.step}, ${task.mockups} mockups, ${task.variants} variants, started ${ago(task.created_at)}${waiting ? `, waiting ${ago(waiting).replace(" ago", "")}` : ""}` : "idle"],
    ["Counters", task ? JSON.stringify(task.metadata) : ""],
    ["Round", data.paths ? `${data.paths.open} open, ${data.paths.used} used, ${data.paths.quarantined} quarantined${Object.keys(data.paths.strikes ?? {}).length ? `, strikes ${Object.entries(data.paths.strikes ?? {}).map(([id, n]) => `${id}x${n}`).join(" ")}` : ""}` : "no paths"],
    ["Live", `${data.live.products} products, ${data.live.expiring} expiring within a day, newest ${ago(data.live.newest)}, oldest ${ago(data.live.oldest)}`],
  ];
  box.innerHTML = `<h2>Automator</h2>${rows(list)}`;
}

function cloud(data: Stats | null) {
  const box = root?.querySelector("[data-cloud]");
  if (!box) return;
  if (!data) {
    box.innerHTML = "<h2>Cloud</h2><p class=\"fine\">No stats file yet.</p>";
    return;
  }
  const lambdas = Object.entries(data.lambdas).map(([name, one]) => `<tr><th>${esc(name)}</th><td>${one.invocations} runs, ${one.errors} errors, ${one.throttles} throttles, ${Math.round(one.duration_ms)} ms avg</td></tr>`);
  const list: [string, unknown][] = [
    ["Stats", `${ago(data.at)}, last ${data.hours} h`],
    ["CDN", data.cdn.requests == null ? "no distribution" : `${data.cdn.requests} requests, ${bytes(data.cdn.bytes)}, ${data.cdn.error_4xx}% 4xx, ${data.cdn.error_5xx}% 5xx`],
    ["Bucket", `${data.bucket.site_objects} site files (${bytes(data.bucket.site_bytes)}), ${data.bucket.data_objects} data files (${bytes(data.bucket.data_bytes)})`],
    ["Shop", `${data.bucket.products} live products, ${data.bucket.designs} designs on the CDN, ${data.bucket.posts} posts`],
  ];
  box.innerHTML = `<h2>Cloud</h2>${rows(list)}<table>${lambdas.join("")}</table>`;
}

function errors(data: Stats | null) {
  const box = root?.querySelector("[data-errors]");
  if (!box) return;
  const found = Object.entries(data?.errors ?? {}).flatMap(([name, list]) => list.map((one) => `${one.at ?? "sometime"} ${one.level} ${name} ${one.message}`));
  box.innerHTML = `<h2>Errors</h2>${found.length ? code(found) : "<p class=\"fine\">None in the window.</p>"}`;
}

function log(data: Automator | null) {
  const box = root?.querySelector("[data-log]");
  if (!box) return;
  box.innerHTML = `<h2>Log</h2>${data?.log?.length ? code(data.log.slice(-120)) : "<p class=\"fine\">No log yet.</p>"}`;
}

async function refresh() {
  const [auto, stats] = await Promise.all([fetchJson<Automator>("/status/automator.json"), fetchJson<Stats>("/status/stats.json")]);
  automator(auto);
  cloud(stats);
  errors(stats);
  log(auto);
}

if (root) {
  refresh();
  setInterval(refresh, EVERY);
}
